src/proofs/range_proof_wip.rs

```diff
@@ -0,0 +1,571 @@
+#![allow(non_snake_case)]
+
+/*
+
+Copyright 2018 by Kzen Networks
+
+This file is part of bulletproof library
+(https://github.com/KZen-networks/bulletproof)
+
+bulletproof is free software: you can redistribute
+it and/or modify it under the terms of the GNU General Public
+License as published by the Free Software Foundation, either
+version 3 of the License, or (at your option) any later version.
+
+@license GPL-3.0+ <https://github.com/KZen-networks/bulletproof/blob/master/LICENSE>
+*/
+
+// based on the paper: https://eprint.iacr.org/2020/735.pdf
+//
+// Bulletproofs (https://eprint.iacr.org/2017/1066) uses the inner product argument.
+// Bulletproofs+ (https://eprint.iacr.org/2020/735.pdf) uses the weighted inner product argument
+// which reduces the overall prover communication by ~15%
+// 
+use curv::arithmetic::traits::{Converter, Modulo};
+use curv::cryptographic_primitives::hashing::hash_sha256::HSha256;
+use curv::cryptographic_primitives::hashing::traits::*;
+use curv::elliptic::curves::traits::*;
+use curv::BigInt;
+use curv::{FE, GE};
+use itertools::iterate;
+use proofs::weighted_inner_product::WeightedInnerProdArg;
+use std::ops::{Shl, Shr};
+use Errors::{self, RangeProofError};
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct RangeProofWIP {
```

Comment:
consider adding  a statement struct that includes all the g_vec,h_vec,G,H etc... This way the verifier takes a proof and a statement and produce 0/1 .

Reply:
Done. I have not added the `ped_com` (commitments to amounts) in the struct as they need to be generated only by prover using secrets. I hope that is okay.

Reply:
What I am usually doing is defining a struct for Witness struct for Statement and struct for Proof. 
prover gets as input Witness and Statement and output a Proof
verifier : is either a function of the Proof . Or gets Statement and Proof and outputs Result